### PR TITLE
Vertical funnel breakdowns

### LINKED ISSUE #5428: Allow removing / adding breakdown values in Funnels UI
## Bug description

Now that we're consistent with breakdown values: https://github.com/PostHog/posthog/issues/5341, 

A new weirdness appears: https://app.posthog.com/insights?insight=FUNNELS&properties=%5B%5D&filter_test_accounts=true&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%7D%5D&actions=%5B%5D&interval=day&new_entity=%5B%5D&funnel_viz_type=steps&display=FunnelViz&date_from=dStart&breakdown=%24device&breakdown_type=event

Say, I want to analyse breakdowns, but most of these values are `None`:

image

Would be nice if I can just remove this from the graph to focus on the rest, like we allow in Trends.

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Implementation of @clarkus's mockups in [#103](https://github.com/PostHog/product-internal/issues/103). Note that this PR significantly changes how our funnel step viz's look by combining the step bars with the table.

- [x] Ability to toggle breakdowns in table
- [x] Percentages of counts in y-axis
- [x] Edge cases in math used to construct bars
- [ ] Toggle to display percentages/absolute values (Punting to separate PR)
- [ ] Option for conversion rates visualization (Punting to separate PR)
- [x] Bar on hover and click
- [x] Table as legend

Note this branch merges in changes from #5730 because I'm anticipating changes to this viz will require the full window width. Merging this PR is blocked on #5730. Please refer to mockups in  [#103](https://github.com/PostHog/product-internal/issues/103) for more information.

Closes [#103](https://github.com/PostHog/product-internal/issues/103). 
Closes #5428.

**Before**

https://user-images.githubusercontent.com/13460330/133555567-03a0eb4c-a40f-4702-9120-f6b738335d8b.mov

**Adding/removing steps and different breakdowns**

https://user-images.githubusercontent.com/13460330/133554892-853780be-509b-48e8-92d0-eb811e42d250.mov

**Opening persons modal in different steps (+ baseline) and toggling legend**

https://user-images.githubusercontent.com/13460330/133554089-a89e6ea2-d3ff-4f2b-a53a-64242049ad04.mov

**Smoke regression test - vertical and horizontal layouts x w/ and w/o breakdowns**

https://user-images.githubusercontent.com/13460330/133555036-c933ef1f-f8e4-4c74-9707-469ca86e43bd.mov

**Infamous $current_url breakdown**

https://user-images.githubusercontent.com/13460330/133560730-861e2c4c-3c4b-48a4-ab9e-cee8c027f1e4.mov

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by clarkus** [comment]
For the ghost bars, I think we can drop the patterned background for now. I was going to reserve those for comparison ranges at some point in the future. 

Screen Shot 2021-08-25 at 9 40 12 AM

In the examples above, is the white background around the third step's bars intentional?

**Comment by alexkim205** [comment]
> In the examples above, is the white background around the third step's bars intentional?

Good call, I had the contrast on my monitor setup so high that I didn't even notice this white bar. :'(

**Comment by mariusandra** [comment]
What I like about our current breakdown graphs is that they give a sense of scale, for example here:

image

versus the new implementation showing breakdown through percentages:

image

Without seeing the table below, you don't know that the light blue box is 91 events, but the reddish one is just 3 events. You'd think Chrome and Mobile Firefox have the same number of users.

If we want to nail diagnosis, we should nail exploration, and looking at the same data through different lenses can yield invaluable insights. 

Thus what about keeping the current breakdown visualisation in addition to the new one? I'd even suggest adding even more ways to break down data, like donuts-inside-donuts, and so on.

image

All of these could just go into this menu, as they're showing the different data in a different way:

image

**Comment by clarkus** [comment]
My concern is that "left-to-right" and "top-to-bottom" are insufficient labels to describe the differences between layouts. I honestly don't think we need alternate layouts for funnels... that was mostly an attempt to satisfy users who were more attached to one layout over another for historical / nostalgia reasons. 

If we are going to offer different rendering modes, we need to focus on that difference in our dropdown option labels. Something like "visualize absolute changes in value" versus "visualize normalized changes in value" or something. 

@kpthatsme you and I have chatted quite a bit about funnels, so I wanted to mention you in case you had any thoughts on this.

**Review by paolodamico — COMMENTED** [review]
Product-wise looking good so far! Tag me for a full review when ready 😉

**Comment by kpthatsme** [comment]
> What I like about our current breakdown graphs is that they give a sense of scale, for example here:
> 
> 
> 
> image
> 
> 
> 
> versus the new implementation showing breakdown through percentages:
> 
> 
> 
> image
> 
> 
> 
> Without seeing the table below, you don't know that the light blue box is 91 events, but the reddish one is just 3 events. You'd think Chrome and Mobile Firefox have the same number of users.
> 
> 
> 
> If we want to nail diagnosis, we should nail exploration, and looking at the same data through different lenses can yield invaluable insights. 
> 
> 
> 
> Thus what about keeping the current breakdown visualisation in addition to the new one? I'd even suggest adding even more ways to break down data, like donuts-inside-donuts, and so on.
> 
> 
> 
> image
> 
> 
> 
> All of these could just go into this menu, as they're showing the different data in a different way:
> 
> 
> 
> image
> 
> 

We should have absolute numbers and maybe percentages visible in each bar - I forget where it is but I'm pretty sure that @clarkus has a mock-up with absolute numbers in the bars - Chris does such a thing exist or am I making it up?

In general the primary point of viewing a funnel is to understand conversion rates - absolute numbers are a secondary metric when viewing a funnel. 

I agree they should absolutely be part of it, but to orient the design around that use case wholly I think would be the wrong move, as the tool focus on understanding absolute numbers is actually a trend chart, not a funnel chart. For example in funnels we're never going to show the absolute number of actual conversions, since a user can only ever appear once. 

However I think this is will be a non issue because we should have these numbers easily available in this view too.

**Comment by clarkus** [comment]
> We should have absolute numbers and maybe percentages visible in each bar - I forget where it is but I'm pretty sure that @clarkus has a mock-up with absolute numbers in the bars - Chris does such a thing exist or am I making it up?

Yup that exists, but those values limit our ability to show more breakdown options in each step. So if the breakdown / grouping has a lot of options, we'd either constrain the number shown concurrently, or drop the absolute numbers in favor of showing absolutes on hover. 

Screen Shot 2021-08-26 at 1 25 49 PM

**Comment by kpthatsme** [comment]
> > We should have absolute numbers and maybe percentages visible in each bar - I forget where it is but I'm pretty sure that @clarkus has a mock-up with absolute numbers in the bars - Chris does such a thing exist or am I making it up?
> 
> 
> 
> Yup that exists, but those values limit our ability to show more breakdown options in each step. So if the breakdown / grouping has a lot of options, we'd either constrain the number shown concurrently, or drop the absolute numbers in favor of showing absolutes on hover. 
> 
> 
> 
> Screen Shot 2021-08-26 at 1 25 49 PM
> 
> 

The intention of this (LTR property breakdown) is to be able to take away all the key information from first glance at the funnel - so I don't think the hover approach is going to solve that pain point. 

What about a horizontal scroll in the scenario you are talking about where there are too many properties? 

In theory I think we'd need to support such behavior on funnels anyway and IMO it's acceptable and puts the control in the users hand for what they want to see (they can restrict their funnel events to only a few properties, for example - or eventually use the table/legend to show/hide).

**Comment by alexkim205** [comment]
> What about a horizontal scroll in the scenario you are talking about where there are too many properties?

> In theory I think we'd need to support such behavior on funnels anyway and IMO it's acceptable and puts the control in the users hand for what they want to see (they can restrict their funnel events to only a few properties, for example - or eventually use the table/legend to show/hide).

This is a good point. I can see it being difficult to determine which bar color corresponds to which absolute value in the table if we only show it on hover. My only concern is if we have a ton of breakdown properties (~20 properties per step), an entire step would take up the full width of the window by default. Perhaps we can limit the number of bars in each step to 5 (or 6? or 7?) and leave the rest unchecked by default so that the user isn't overwhelmed by a ton of breakdown options from the first query.

**Comment by alexkim205** [comment]
Added table legend check box logic


https://user-images.githubusercontent.com/13460330/131085332-57601968-f6a8-43fe-9d9d-8703dad0ef71.mov

**Comment by clarkus** [comment]
+1 on giving users more control here. We need to identify upper and lower bounds for the count of bars in each step. There is going to be some upper bound where the visualization breaks down. I was working on this, but wound up moving on to other hot topics. I can revisit this week. 

Other considerations might include showing baseline conversion. There are cases where a user might want to compare a single value in a breakdown group to the baseline conversion of the funnel.

**Comment by Twixes** [comment]
Will wait for Alex to handle next week.

**Comment by alexkim205** [comment]
Update - adding table as a legend in this PR

Screen Shot 2021-09-14 at 7 46 21 PM

**Comment by alexkim205** [comment]
There's still a few rough edges to work out but the new visualization functionally works

Screen Shot 2021-09-15 at 4 00 49 PM

**Comment by alexkim205** [comment]
Finally ready for review! Tagging @mariusandra @paolodamico @clarkus @kpthatsme for a 👀

**Comment by marcushyett-ph** [comment]
Wow this is looking epic @alexkim205 - I'm curious as to how we're going to annotate the bars (I see both no annotation and numeric - but not sure if that makes sense if we're rendering %age)?

**Review by mariusandra — DISMISSED** [review]
1. This seems really cool, but also like a pretty big difference. There's no way I'm able to review that all the numbers in the table/chart/etc are perfect with every combination of data. Plus it might also be a subjective change that should get a bit more feedback. 

Thus, an annoying request: can you add all the new changes behind a feature flag? Feel free to use new features flags view in the toolbar for magic :).

2. The visibility of the bars is off by default. They should always show.



3. Introducing yet another local state (local = in kea, but not synced to the URL/filters) for this visibility seems like introducing tech debt though. Any way to hook it up to the URL... and before making an API request to see if it's needed?

4. If I breakdown by something with just 1 prop, I'd expect these lines to be different. Isn't "baseline" just null?

image

5. I kind of expect nothing to show in the funnel after this:



(not sure if this is scope creep?)

**Comment by alexkim205** [comment]
> Wow this is looking epic @alexkim205 - I'm curious as to how we're going to annotate the bars (I see both no annotation and numeric - but not sure if that makes sense if we're rendering %age)?

Thanks! Right now numeric counts will be shown above each bar if we're looking at 5 or less breakdown bars at once and if there's any more, I have it so that labels are hidden right now. I think it was brought up previously that having the numbers and percentages in the table align with each step would be a good enough visual indicator for dropoff. I feel like having extra annotation might get redundant, but maybe the table data isn't clear enough or should be highlighted better?

**Comment by marcushyett-ph** [comment]
All clear to me - I was just curious the rationale makes a lot of sense

**Comment by alexkim205** [comment]
Thanks Marius for the thorough review :)

1. Nice FF's in toolbar makes it a lot easier to do that. Done!
2. fixed
3. stored state in URL instead of separate kea value
4. From my understanding and how I implemented it, baseline is supposed to represent the total of the breakdown bars. If there's only one breakdown property however, I think it makes sense to hide baseline. 
5. fixed

**Comment by mariusandra** [comment]
Looks cool, and it's definitely fun to enable/disable the feature with the toolbar. Some more things:

1. With the flag off:

image

... it still looks different than `master`:

image

2. The visibility filter itself it broken:



My suggestions:
- instead of a visibility map, convert it to a hidden map. This way everything is visible by default, and `visible == !hidden[key]`, instead of `visible == visible[key] === undefined || visible[key]`.
- Index the visibility by key, and let's define a schema or it'll get out of hand quickly. If you index by index and the breakdown order changes from one day to the next, it'll fail. 
- Regarding the schema, something like `hidden = { "event/$pageview/${breakdownValue}": true, "action/1": true }` is probably enough. 
- Keep in mind this will also be used in the future to save the state of the retention toggles, and who knows what else.

**Review by clarkus — APPROVED** [review]
These changes look good to me overall. The only thing I think we should consider changing is the label for the funnel layout control. Top-to-bottom and Left-to-right aren't very descriptive of the layouts now. Top to bottom is visualizing breakdowns using stacked bars, but left to right is using distinct bars. I'm trying to think of alternate labels, but coming up blank. Probably fine to ship with these labels, but I feel like we could come up with something better. I will keep thinking about it and file a follow up issue.

**Comment by paolodamico** [comment]
Looks like @mariusandra has reviewed the code and @clarkus the UI/UX, so withdrawing from review to avoid duplicated effort.

**Comment by alexkim205** [comment]
The hidden map concept is pretty sleek, thanks @mariusandra. Recent changes include: 

- Pushed up fixes to use a hidden map instead
- refactored and standardized visibility indices
- fixed a time conversion bug
- few more styling bugs that broke w/ merge conflicts

File: frontend/src/scenes/funnels/funnelLogic.ts [hunk 1]
```diff
@@ -80,6 +83,7 @@ export const cleanFunnelParams = (filters: Partial<FilterType>, discardFiltersNo
             : {}),
         ...(filters.funnel_window_interval ? { funnel_window_interval: filters.funnel_window_interval } : {}),
         ...(filters.funnel_order_type ? { funnel_order_type: filters.funnel_order_type } : {}),
+        ...(filters.hidden_map ? { hidden_map: filters.hidden_map } : {}),
```
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:86]: Nit that won't hold up the PR: could we find a name that better describes what this does? The `map` here doesn't really say what is hidden. Even `hiddenLines` would be more descriptive, if not slightly misleading...

File: frontend/src/scenes/funnels/funnelLogic.ts [hunk 2]
```diff
@@ -566,6 +673,16 @@ export const funnelLogic = kea<funnelLogicType>({
 
     listeners: ({ actions, values, props }) => ({
         loadResultsSuccess: async () => {
+            // set visibility of baseline or first five breakdowns for each step
+            values.steps?.forEach((step) => {
+                values.flattenedStepsByBreakdown
+                    ?.filter((s) => !!s.breakdown)
+                    ?.slice(5)
+                    .forEach((b) => {
+                        actions.setHiddenById({ [getVisibilityIndex(step, b.breakdown_value)]: true })
```
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:682]: The comment is slightly misleading, as this sets the visibility of everything except the first 5?

**Review by mariusandra — APPROVED** [review]
It's feature flagged, so let's get it in and start testing!

Things to fix in a follow up PR:
- Small nit in the code about the hidden map
- This seems broken, possibly thanks to the hidden map.

File: frontend/src/scenes/funnels/funnelLogic.ts (hunk 2 above)
Comment (alexkim205) [reply frontend/src/scenes/funnels/funnelLogic.ts:689]: got it! will update the comment

File: frontend/src/scenes/funnels/funnelLogic.ts (hunk 1 above)
Comment (alexkim205) [reply frontend/src/scenes/funnels/funnelLogic.ts:86]: Hm, would `hiddenLegendKeys` work in this context?

**Comment by alexkim205** [comment]
Follow up PR here #6061.

**Pull request merged**